File: stacks/laravel/app/Models/BillingDetail.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

/**
 * BillingDetail — 会計の明細。
 *
 * unit_price が null の行は、金額の合計に**0円として含めてはならない**
 * （spec/model.md・検算2）。App\Services\BillingCalculator が絶対にこの規則を守る。
 */
class BillingDetail extends Model
{
    protected $fillable = ['billing_id', 'row_no', 'price_code', 'name', 'quantity', 'unit_price', 'is_taxable'];

    protected $casts = [
        'quantity' => 'decimal:2',
        'is_taxable' => 'boolean',
    ];

    public function billing(): BelongsTo
    {
        return $this->belongsTo(Billing::class);
    }

    public function hasPrice(): bool
    {
        return $this->unit_price !== null;
    }
}

File: stacks/rails/app/models/billing_detail.rb
# 会計の明細。unit_price が未設定（null）の行がありうる。
# **0円として合計に入れてはならない**（spec/model.md 13章・検算2）。
class BillingDetail < ApplicationRecord
  belongs_to :billing

  validates :price_code, presence: true
  validates :name, presence: true
  validates :quantity, numericality: { greater_than: 0 }

  before_validation :assign_row_no, on: :create

  # quantity × unit_price。unit_price が無ければ nil（合計に含めない）。
  def amount
    return nil if unit_price.nil?

    (quantity.to_d * unit_price.to_d)
  end

  private

  def assign_row_no
    return if row_no.present?
    return if billing.nil?

    self.row_no = BillingDetail.unscoped.where(billing_id: billing_id).maximum(:row_no).to_i + 1
  end
end
